### DISCUSSION TITLE
Read Data in Shiny

### DISCUSSION THREAD

**Discussion by yohanespm77** [teamalgoritma/community#631] [Q&A] [answered]
Saya punya kodingan seperti ini di Shiny Server : 
<img width="355" alt="image" src="https://user-images.githubusercontent.com/106010284/231713009-a613d5d3-08da-4cc8-be71-d49e1369d312.png">

<img width="668" alt="image" src="https://user-images.githubusercontent.com/106010284/231713115-4053e083-d58c-4bc9-9354-eb747e7538c4.png">

<img width="511" alt="image" src="https://user-images.githubusercontent.com/106010284/231713195-076c0f12-24dc-474e-b913-1c3d5aee4cff.png">

Pertanyaan saya, bagaimana caranya agar datanya dibaca hanya sekali saja, tidak berulang-ulang seperti kodingan saya ?

Mohon bantuannya kak, terima kasih 🙏

**Comment by IrfanChairurrachman** [comment, marked as answer]
Dear @yohanespm77 ,

Dari kodingan di atas, Anda mencoba untuk membuat fitur upload file dimana file tersebut akan ditampilkan dalam bentuk Table dan juga Plotly. Pada kodingan di atas, input file harus dipanggil berulang-ulang di dalam masing-masing fungsi output. 

Untuk membuat kode tidak redundan, Anda dapat assign input file ke dalam sebuah objek di luar fungsi output (namun tetap di dalam fungsi server), sehingga ketika Anda membutuhkan data yang diupload tersebut, Anda langsung memanggil nama objeknya saja. Contoh penerapannya seperti berikut:
https://mastering-shiny.org/action-transfer.html#uploading-data 

Best practicenya, untuk menyimpan data input ke dalam sebuah objek, sebaiknya di balut ke dalam fungsi `reactive`, dimana nilainya hanya akan berubah ketika terdapat perubahan/trigger dari user. Fungsi reactive ini telah digunakan pada contoh kode referensi di atas. Namun, jika Anda ingin membaca lebih lanjut mengenai `reactive` dapat mengacu pada dokumentasi berikut: https://rdrr.io/cran/shiny/man/reactive.html.

Sekian, semoga membantu.
